experimental/wazerotest/wazerotest.go

```diff
@@ -653,38 +653,53 @@ type StackFrame struct {
 	SourceOffset uint64
 }
 
-type stackIterator struct {
+// StackIterator is an implementation of the experimental.StackIterator
+// interface.
+type StackIterator struct {
+	index int
 	stack []StackFrame
 	fndef []api.FunctionDefinition
-	index int
 }
 
-func (si *stackIterator) Next() bool {
+// Next advances the iterator to the next stack frame.
+func (si *StackIterator) Next() bool {
 	si.index++
 	return si.index < len(si.stack)
 }
 
-func (si *stackIterator) FunctionDefinition() api.FunctionDefinition {
+// FunctionDefinition returns the definition for the function recorded on the
+// current stack frame.
+func (si *StackIterator) FunctionDefinition() api.FunctionDefinition {
 	return si.fndef[si.index]
 }
 
-func (si *stackIterator) Parameters() []uint64 {
-	return si.stack[si.index].Params
+// SourceOffset returns the offset in the source code of the current stack
+// frame.
+func (si *StackIterator) SourceOffset() uint64 {
+	return si.stack[si.index].SourceOffset
 }
 
-func (si *stackIterator) SourceOffset() uint64 {
-	return si.stack[si.index].SourceOffset
+// Parameters returns the parameters of the current stack frame.
+func (si *StackIterator) Parameters() []uint64 {
+	return si.stack[si.index].Params
 }
 
-func (si *stackIterator) reset() {
-	si.index = 0
+// Reset positions the stack iterator back on the top most frame.
+func (si *StackIterator) Reset() {
+	si.index = -1
 }
 
-func newStackIterator(stack []StackFrame) *stackIterator {
-	si := &stackIterator{
-		stack: stack,
+// NewStackIterator constructs a stack iterator from a list of stack frames.
+// The top most frame is the last one.
+func NewStackIterator(stack ...StackFrame) *StackIterator {
```

Comment:
this made me realize we have exposed functions that are experimental in wazerotest. We have to make sure that when wazerotest moves out of experimental we scrub for things like this. I don't think anyone can depend on wazerotest meanwhile because it will drift even if a lot of it is stable (based on api package types)

Reply:
That's a good point.

The only experimental thing currently exposed in wazerotest is the benchmark function: https://pkg.go.dev/github.com/tetratelabs/wazero@v1.1.1-0.20230509001827-ed17e523ba21/experimental/wazerotest#BenchmarkFunctionListener

Let's move this function up to `experimental`, as well as the `StackIterator` type here?

Reply:
I went ahead and moved all dependencies on `experimental` from `wazerotest` in https://github.com/tetratelabs/wazero/pull/1444/commits/979f17735d1a04e4931950fd1eb8f4e8be69fc19